web/app/js/components/Top.jsx

```diff
@@ -0,0 +1,363 @@
+import { defaultMaxRps, processTapEvent } from './util/TapUtils.jsx';
+import './../../css/tap.css';
+
+class Top extends React.Component {
+  static propTypes = {
+    api: PropTypes.shape({
+      PrefixedLink: PropTypes.func.isRequired,
+    }).isRequired,
+    pathPrefix: PropTypes.string.isRequired
+  }
+
+  constructor(props) {
+    super(props);
+    this.api = this.props.api;
+    this.loadFromServer = this.loadFromServer.bind(this);
+
+    this.state = {
+      tapResultsById: {},
+      topEventIndex: {},
+      error: null,
+      resourcesByNs: {},
+      authoritiesByNs: {},
+      query: {
+        resource: "",
+        namespace: "",
+        toResource: "",
+        toNamespace: "",
+        method: "",
+        path: "",
+        scheme: "",
+        authority: "",
+        maxRps: defaultMaxRps
+      },
+      maxRowsToStore: 40,
+      awaitingWebSocketConnection: false,
+      tapRequestInProgress: false,
+      pollingInterval: 10000,
+      pendingRequests: false
+    };
+  }
+
+  componentDidMount() {
+    this.startServerPolling();
+  }
+
+  componentWillUnmount() {
+    if (this.ws) {
+      this.ws.close(1000);
+    }
+    this.stopTapStreaming();
+    this.stopServerPolling();
+  }
+
+  onWebsocketOpen = () => {
+    let query = _.cloneDeep(this.state.query);
+    query.maxRps = parseFloat(query.maxRps);
+
+    this.ws.send(JSON.stringify({
+      id: "top-web",
+      ...query
+    }));
+    this.setState({
+      awaitingWebSocketConnection: false,
+      error: null
+    });
+  }
+
+  onWebsocketRecv = e => {
+    this.indexTapResult(e.data);
+  }
+
+  onWebsocketClose = e => {
+    this.stopTapStreaming();
+
+    if (!e.wasClean) {
+      this.setState({
+        error: {
+          error: `Websocket [${e.code}] ${e.reason}`
+        }
+      });
+    }
+  }
+
+  onWebsocketError = e => {
+    this.setState({
+      error: { error: e.message }
+    });
+
+    this.stopTapStreaming();
+  }
+
+  getResourcesByNs(rsp) {
+    let statTables = _.get(rsp, [0, "ok", "statTables"]);
+    let authoritiesByNs = {};
+    let resourcesByNs = _.reduce(statTables, (mem, table) => {
+      _.map(table.podGroup.rows, row => {
+        if (!mem[row.resource.namespace]) {
+          mem[row.resource.namespace] = [];
+          authoritiesByNs[row.resource.namespace] = [];
+        }
+
+        switch (row.resource.type.toLowerCase()) {
+          case "service":
+            break;
+          case "authority":
+            authoritiesByNs[row.resource.namespace].push(row.resource.name);
+            break;
+          default:
+            mem[row.resource.namespace].push(`${row.resource.type}/${row.resource.name}`);
+        }
+      });
+      return mem;
+    }, {});
+    return {
+      authoritiesByNs,
+      resourcesByNs
+    };
+  }
+
+  parseTapResult = data => {
+    let d = processTapEvent(data);
+
+    if (d.eventType === "responseInit") {
+      d.success = parseInt(d.http.responseInit.httpStatus, 10) < 500;
```

Comment:
This unfortunately doesn't capture the success rate properly for failed gRPC requests, since even gRPC failures are returned with HTTP status 200. In the CLI implementation we're using:

```
	success := req.rspInit.GetHttpStatus() < 500
	if success {
		switch eos := req.rspEnd.GetEos().GetEnd().(type) {
		case *pb.Eos_GrpcStatusCode:
			success = eos.GrpcStatusCode == 0

		case *pb.Eos_ResetErrorCode:
			success = false
		}
	}
```

And I think we need to do something similar here.

Reply:
ahhh good spot, thanks!